File: Python/Planners/Curvesbased/PlanPRMPath.py
from math import ceil, fabs
from cv2 import waitKey
import numpy as np
import sortedcontainers
import globalvar
import cv2
from TransformPathToTrajectory import TransformPathToTrajectory
from main_unstructure import checkObj_linev, checkObj_point, CreateVehiclePolygon, distance, \
    GenerateStaticObstacles_unstructured, VisualizeStaticResults, VisualizeDynamicResults


def calculate_xy_to_index(x,y):
    idx_x = ceil((x-globalvar.planning_scale_.xmin)/globalvar.resolution_x)
    idx_y = ceil((y-globalvar.planning_scale_.ymin)/globalvar.resolution_y)
    if idx_x < 0:
        idx_x = 0
    if idx_x >= globalvar.num_nodes_x:
        idx_x = globalvar.num_nodes_x - 1
    if idx_y < 0:
        idx_y = 0
    if idx_y >= globalvar.num_nodes_y:
        idx_y = globalvar.num_nodes_y - 1
    return idx_x,idx_y


def CreateCostmap():
    cost_ndarr = np.zeros((globalvar.num_nodes_x,globalvar.num_nodes_y,1),dtype=np.uint8)
    fill_list = []
    for obstacle in globalvar.obstacles_[0]:
        obs = np.vstack((obstacle.x, obstacle.y))
        obs_idx = []
        for col in range(obs.shape[1]):
            idx_x,idx_y = calculate_xy_to_index(obs[0,col],obs[1,col])
            obs_idx.append((idx_x,idx_y))
        fill_list.append(np.array(obs_idx))
    cv2.fillPoly(cost_ndarr,np.array(fill_list),255)
    

    length_unit = 0.5*(globalvar.resolution_x+globalvar.resolution_y)
    radius = globalvar.vehicle_geometrics_.radius/length_unit
    radius = int(radius)
    elemEllipse = cv2.getStructuringElement(cv2.MORPH_ELLIPSE,(radius,radius))
    costmap_ = cv2.dilate(cost_ndarr,elemEllipse)
    return costmap_



def IsVertexValid(x,y):
    x_inf = globalvar.planning_scale_.xmin
    x_sup = globalvar.planning_scale_.xmax
    y_inf = globalvar.planning_scale_.ymin
    y_sup = globalvar.planning_scale_.ymax
    if x> x_sup or x<x_inf or y>y_sup or y<y_inf:
        return False
    idx_x, idx_y = calculate_xy_to_index(x,y)
    if globalvar.costmap_[idx_x][idx_y] == 255:
        return False
    return True

def Is3DNodeValid(x,y,theta):
    xr = x + globalvar.vehicle_geometrics_.r2x * np.cos(theta)
    yr = y + globalvar.vehicle_geometrics_.r2x * np.sin(theta)
    xf = x + globalvar.vehicle_geometrics_.f2x * np.cos(theta)
    yf = y + globalvar.vehicle_geometrics_.f2x * np.sin(theta)
    x_inf = globalvar.planning_scale_.xmin + globalvar.vehicle_geometrics_.radius * 1.01
    x_sup = globalvar.planning_scale_.xmax - globalvar.vehicle_geometrics_.radius * 1.01
    y_inf = globalvar.planning_scale_.ymin + globalvar.vehicle_geometrics_.radius * 1.01
    y_sup = globalvar.planning_scale_.ymax - globalvar.vehicle_geometrics_.radius * 1.01
    if xr > x_sup or xr < x_inf or xf > x_sup or xf < x_inf or yr > y_sup or yr < y_inf or yf > y_sup or yf < y_inf:
        return False
    idx_r_x, idx_r_y = calculate_xy_to_index(xr,yr)
    idx_f_x, idx_f_y = calculate_xy_to_index(xf,yf)
    if globalvar.costmap_[idx_r_y,idx_r_x]==255 or globalvar.costmap_[idx_f_y,idx_f_x]==255:
        return False
    return True


# check feasible edge when produced
def IsLineValid(p1, p2):  # p1(x,y),p2(x,y) represent a vector p1----->p2
    theta = np.arctan2(p2[1] - p1[1], p2[0] - p1[0])
    V1 = CreateVehiclePolygon(x=p1[0], y=p1[1], theta=theta)
    V2 = CreateVehiclePolygon(x=p2[0], y=p2[1], theta=theta)
    X = np.hstack((V1.x, V2.x))
    Y = np.hstack((V1.y, V2.y))
    min_x_index = np.argmin(X)
    max_x_index = np.argmax(X)
    min_y_index = np.argmin(Y)
    max_y_index = np.argmax(Y)
    min_x_v = np.min(X)
    max_x_v = np.max(X)
    min_y_v = np.min(Y)
    max_y_v = np.max(Y)
    P1 = np.array([X[min_x_index], Y[min_x_index]])
    P2 = np.array([X[max_x_index], Y[max_x_index]])
    P3 = np.array([X[min_y_index], Y[min_y_index]])
    P4 = np.array([X[max_y_index], Y[max_y_index]])
    for obstacle in globalvar.obstacles_[0]:

        obs = np.vstack((obstacle.x, obstacle.y))
        min_x_obs = np.min(obstacle.x)
        max_x_obs = np.max(obstacle.x)
        min_y_obs = np.min(obstacle.y)
        max_y_obs = np.max(obstacle.y)
        if (min_x_obs >= min_x_v and max_x_obs <= max_x_v and min_y_obs >= min_y_v and max_y_obs <= max_y_v) or \
                (min_x_obs <= min_x_v and max_x_obs >= max_x_v and min_y_obs <= min_y_v and max_y_obs >= max_y_v):
            return False
        if checkObj_linev(P1, P4, obs) or checkObj_linev(P4, P2, obs) or checkObj_linev(P2, P3, obs) or checkObj_linev(
                P3, P1, obs):
            return False

    return True


def getvertex(k):
    vertex = np.array([globalvar.vehicle_TPBV_.x0, globalvar.vehicle_TPBV_.y0])
    vertex = np.vstack((vertex, [globalvar.vehicle_TPBV_.xtf, globalvar.vehicle_TPBV_.ytf]))
    lx = globalvar.planning_scale_.xmin
    ux = globalvar.planning_scale_.xmax
    ly = globalvar.planning_scale_.ymin
    uy = globalvar.planning_scale_.ymax

    while vertex.shape[0] < k + 2:
        x = (ux - lx) * np.random.rand() + lx
        y = (uy - ly) * np.random.rand() + ly
        pt = np.array([x, y])
        if not IsVertexValid(x,y):
            continue
        vertex = np.vstack((vertex, pt))
    return vertex


def get_edges(vertex):
    size = vertex.shape[0]
    edges = []
    for i in range(size):
        edges.append([])
    for i in range(size):
        for j in range(i + 1, size):
            p1 = vertex[i, :]
            p2 = vertex[j, :]
            dist = np.linalg.norm(p1-p2)
            x_samp = np.linspace(p1[0],p2[0],num=ceil(dist/(0.1*globalvar.vehicle_geometrics_.radius)))
            y_samp = np.linspace(p1[1],p2[1],num=ceil(dist/(0.1*globalvar.vehicle_geometrics_.radius)))
            theta_ = np.arctan2(p2[1]-p1[1],p2[0]-p1[0])
            IsfeasibleLine = 1
            for x, y in zip(x_samp,y_samp):
                if not Is3DNodeValid(x,y,theta_):
                    IsfeasibleLine = 0
                    break
            if IsfeasibleLine:
                edges[i].append(j)
                edges[j].append(i)
    return edges


class PriorityQueue(object):

    def __init__(self, node):
        self._queue = sortedcontainers.SortedList([node])

    def push(self, node):
        self._queue.add(node)

    def pop(self):
        return self._queue.pop(index=0)

    def empty(self):
        return len(self._queue) == 0

    def compare_and_replace(self, i, node):
        if node < self._queue[i]:
            self._queue.pop(index=i)
            self._queue.add(node)

    def find(self, node):
        try:
            loc = self._queue.index(node)
            return loc
        except ValueError:
            return None


class Node(object):
    def __init__(self, vertex_order=0, g=0, h=0, parent=None):
        self.vertex_order = vertex_order
        self.g = g
        self.h = h
        self.f = g + h
        self.parent = parent

    def __lt__(self, other):
        return self.f < other.f

    def __eq__(self, other):
        return self.vertex_order == other.vertex_order


def resample(path):
    for i in range(len(path) - 1):
        dis = distance(path[i], path[i + 1])
        if i == 0:
            obj_x = np.linspace(path[i][0], path[i + 1][0], round(dis))
            obj_x = np.delete(obj_x, -1)
            obj_y = np.linspace(path[i][1], path[i + 1][1], round(dis))
            obj_y = np.delete(obj_y, -1)
        elif i == len(path) - 2:
            obj_x = np.hstack((obj_x, np.linspace(path[i][0], path[i + 1][0], round(dis))))
            obj_y = np.hstack((obj_y, np.linspace(path[i][1], path[i + 1][1], round(dis))))
        else:
            obj_x = np.hstack((obj_x, np.linspace(path[i][0], path[i + 1][0], round(dis))))
            obj_x = np.delete(obj_x, -1)
            obj_y = np.hstack((obj_y, np.linspace(path[i][1], path[i + 1][1], round(dis))))
            obj_y = np.delete(obj_y, -1)
    path = np.vstack((obj_x, obj_y)).T
    return path


def PRM():
    vertex = getvertex(100)

    print('get vertex over')
    edges = get_edges(vertex)
    print('get edges over')
    node = Node(h=distance(vertex[0], vertex[1]))
    completeness_flag = 0
    path_length = 0
    iter_num = 0
    max_iter = 500
    open_list = PriorityQueue(node)
    closed = set()
    path = []
    prev = None
    while (not open_list.empty()) and (not completeness_flag) and (iter_num < max_iter):
        cur_node = open_list.pop()
        closed.add(cur_node.vertex_order)
        for edge in range(len(edges[cur_node.vertex_order])):
            newVertex = edges[cur_node.vertex_order][edge]
            child_g = cur_node.g + distance(vertex[cur_node.vertex_order], vertex[newVertex])
            child_h = distance(vertex[1], vertex[newVertex])
            child_node = Node(vertex_order=newVertex, g=child_g, h=child_h, parent=cur_node)
            if newVertex not in closed and not open_list.find(child_node):
                prev = child_node
                if child_node.vertex_order == 1:
                    path_length = child_node.f
                    completeness_flag = 1
                    break
                open_list.push(child_node)
        iter_num += 1

    while prev is not None:
        curp = vertex[prev.vertex_order]
        path.append(curp)
        prev = prev.parent
    path = list(reversed(path))
    if completeness_flag:
        path = resample(path)
        print('succeed')
    else:
        print('failed')
        return
    x = path[:, 0]
    y = path[:, 1]
    theta = []
    for i in range(len(path) - 1):
        theta.append(np.arctan2(path[i + 1][1] - path[i][1], path[i + 1][0] - path[i][0]))
    theta.append(0)
    theta = np.array(theta)
    return x, y, theta, path_length, completeness_flag


if __name__ == "__main__":
    globalvar.obstacles_ = GenerateStaticObstacles_unstructured()
    globalvar.costmap_  = CreateCostmap()

    [x, y, theta, path_length, completeness_flag] = PRM()
    transmethod_flag = 1  # choose different path to trajectory method
    trajectory = TransformPathToTrajectory(x, y, theta, path_length, transmethod_flag)
    VisualizeStaticResults(trajectory)
    VisualizeDynamicResults(trajectory)

File: Cpp/Planners/Curvesbased/src/PlanPRMPath.cpp
#include "PlanPRMPath.h"
#include "mathstruct.h"
#include "TransformP2T.h"
#include <algorithm>
#include <cmath>
#include <queue>
#include <set>
#include <opencv2/opencv.hpp>

cv::Mat CreateCostmap(){
    cv::Mat costmap = cv::Mat::zeros(num_nodes_x,num_nodes_y,CV_8UC1); 
    for (int i = 0; i < Nobs; i++){
        vector<cv::Point> obs_idx; 
        for (int obs_cell = 0; obs_cell<obstacles_[i].size();obs_cell++){
            math::Vec2d obs_idx_temp = calc_xy_index(obstacles_[i][obs_cell]);
            cv::Point pos = cv::Point(obs_idx_temp.x(),obs_idx_temp.y());
            obs_idx.push_back(pos);
        }
        cv::fillPoly(costmap,obs_idx,cv::Scalar(255));
    }
    
    double length_unit = 0.5*(resolution_x+resolution_y);
    double radius = vehicle_geometrics_.radius/length_unit;
    cv::Mat elemEllipse = cv::getStructuringElement(cv::MORPH_ELLIPSE,cv::Size(radius,radius));
    cv::Mat map;
    cv::dilate(costmap,map, elemEllipse);
    return map;
}


vector<math::Vec2d> getvertex(unsigned int k) {
    math::Vec2d source_vertex(vehicle_TPBV_.x0, vehicle_TPBV_.y0);
    math::Vec2d goal_vertex(vehicle_TPBV_.xtf, vehicle_TPBV_.ytf);
    vector<math::Vec2d> vertex;
    vertex.push_back(source_vertex);
    vertex.push_back(goal_vertex);
    try {
        std::string error = "";
        if (!Is3DNodeValid(vehicle_TPBV_.x0, vehicle_TPBV_.y0, vehicle_TPBV_.theta0)) {
            error = "vehicle can not start path correctly";
            throw error;
        }
        if (!Is3DNodeValid(vehicle_TPBV_.xtf, vehicle_TPBV_.ytf, vehicle_TPBV_.thetatf)) {
            error = "vehicle can not end path correctly";
            throw error;
        }
    }
    catch (std::string error) {
        std::cout << error << std::endl;
    }
    double lx = planning_scale_.xmin;
    double ux = planning_scale_.xmax;
    double ly = planning_scale_.ymin;
    double uy = planning_scale_.ymax;

    while (vertex.size() < k + 2) {
        double x = (ux - lx) * rand01() + lx;
        double y = (uy - ly) * rand01() + ly;
        math::Vec2d point2Dim(x, y);
        if (Is2DNodeValid(x, y)) {
            vertex.push_back(point2Dim);
        }
    }
    return vertex;
}

vector<vector<int>> getedges(vector<math::Vec2d> vertex) {
    unsigned int NumofVertex = vertex.size();
    vector<vector<int>> edges(NumofVertex);
    double radius = vehicle_geometrics_.radius;
    for (unsigned int i = 0; i < NumofVertex; i++) {
        for (unsigned int j = i + 1; j < NumofVertex; j++) {
            math::Vec2d p1 = vertex[i];
            math::Vec2d p2 = vertex[j];
            double angle = (p2 - p1).Angle();
            double distance = (p2 - p1).Length();
            bool Isfeasible_Line = 1;
            double step = radius* 0.1;
            for (double sample=0; sample < distance; sample += step){
                math::Vec2d sample_pt = calc_xy_index(p1.x()+sample*cos(angle), p1.y()+sample*sin(angle));
                cv::circle(costmap_, cv::Point(sample_pt.x(),sample_pt.y()), 1, cv::Scalar(250));
                if(costmap_.at<uchar>(sample_pt.y(),sample_pt.x())==255){
                    cv::circle(costmap_, cv::Point(sample_pt.x(),sample_pt.y()), 3, cv::Scalar(250));
                    Isfeasible_Line = 0;
                    break;
                }
            }
            if (Isfeasible_Line)
            {
                edges[i].push_back(j);
                edges[j].push_back(i);
            }

        }
    }
    return edges;
}
pair <struct Trajectory, double> PlanPRMPath( ) {
    unsigned int k = 20;
    cv::Mat costmap = CreateCostmap();
    vector<math::Vec2d> vertex = getvertex(k);
    vector<vector<int>> edges = getedges(vertex);

    priority_queue<AstarNode*> openlist_;
    AstarNode* init_node = new AstarNode(sizeof(AstarNode));
    init_node->g = 0;
    init_node->h = vertex[0].DistanceTo(vertex[1]);
    init_node->f = init_node->g + init_node->h;
    init_node->parent = nullptr;
    init_node->vertex_order = 0;
    openlist_.push(init_node);

    set<int> closed;

    bool completeness_flag = 0;
    double path_length = 0;
    int iter = 0;

    vector<math::Vec2d> path;
    AstarNode* prev = nullptr;
    const int max_iter = 500;
    int iternum = 0;
    while (!openlist_.empty() && (!completeness_flag) && iternum < max_iter) {
        AstarNode* cur_node = openlist_.top();
        openlist_.pop();
        closed.insert(cur_node->vertex_order);
        for (unsigned int edge = 0; edge < edges[cur_node->vertex_order].size(); edge++) {
            int newVertex = edges[cur_node->vertex_order][edge];
            double child_g = cur_node->g + vertex[cur_node->vertex_order].DistanceTo(vertex[newVertex]);
            double child_h = vertex[1].DistanceTo(vertex[newVertex]);
            AstarNode* child_node = new AstarNode(newVertex,child_g,child_h,child_g+child_h);
            child_node->parent = cur_node;
            if (closed.find(newVertex) == closed.end() && !find(openlist_, newVertex)) {
                prev = child_node;
                if (child_node->vertex_order == 1) {
                    path_length = child_node->f;
                    completeness_flag = 1;
                    break;
                }
                openlist_.push(child_node);
            }
        }
        iternum += 1;
    }
    if (completeness_flag){
        std::cout << "PRM succeed!"<< std::endl;
    }
    else{
        std::cout << "PRM failed!"<< std::endl;
    }
    
    //getpath
    while (prev != nullptr) {
        math::Vec2d curp = vertex[prev->vertex_order];
        path.push_back(curp);
        prev = prev->parent;
    }
    // initial vertexes
    // for (int i = 0; i < path.size(); i++) {
    //     std::cout << path[i].x() << ' ' << path[i].y() << std::endl;
    // }
    reverse(path.begin(), path.end());



    //resample
    vector<double> x;
    vector<double> y;
    vector<double> theta;
    if (completeness_flag) {
        path = Resample(path);
    }
    for (unsigned int i = 0; i < path.size() - 1; i++) {
        x.push_back(path[i].x());
        y.push_back(path[i].y());
        double angle = (path[i + 1] - path[i]).Angle();
        theta.push_back(angle);
    }
    x.push_back(path.back().x());
    y.push_back(path.back().y());
    theta.push_back(0);
    pair <struct Trajectory, double> ret = make_pair(TransformPathToTrajectory(x, y, theta, path_length, 1), path_length);
    return ret;
}

vector<math::Vec2d> Resample(vector<math::Vec2d> vec) {
    vector<math::Vec2d> res;
    for (int i = 0; i < vec.size() - 1; i++) {
        double distance = vec[i].DistanceTo(vec[i + 1]);
        double delta_x = (vec[i + 1].x() - vec[i].x()) / distance;
        double delta_y = (vec[i + 1].y() - vec[i].y()) / distance;
        for (int j = 0; j < distance; j++) {
            math::Vec2d a(j * delta_x, j * delta_y);
            res.push_back(vec[i] + a);
        }
    }
    res.push_back(vec.back());
    return res;
}



bool find(priority_queue<AstarNode*> openlist, int vertex_order) {
    while (!openlist.empty()) {
        AstarNode* cur = openlist.top();
        if (cur->vertex_order == vertex_order)
        {
            return true;
        }
        openlist.pop();
    }
    return false;
}

vector<double> linspace(double pos1, double pos2, int n) {
    vector<double> linvec;
    double d = (pos2 - pos1) / (n - 1);
    for (int i = 0; i < n; i++) {
        linvec.push_back(pos1 + i * d);
    }
    return linvec;
}

bool Is2DNodeValid(double x, double y) {
    double x_inf = planning_scale_.xmin;
    double x_sup = planning_scale_.xmax;
    double y_inf = planning_scale_.ymin;
    double y_sup = planning_scale_.ymax;
    if (x > x_sup || x < x_inf ||
        y > y_sup || y < y_inf) {
        return false;
    }
    math::Vec2d idx = calc_xy_index(math::Vec2d(x,y));
    if (costmap_.at<uchar>(idx.y(),idx.x())==255){
        return false;
    }
    return true;
}

bool Is3DNodeValid(double x, double y, double theta) {
    double xr = x + vehicle_geometrics_.r2x * cos(theta);
    double a = std::cos(theta);
    double yr = y + vehicle_geometrics_.r2x * sin(theta);
    double xf = x + vehicle_geometrics_.f2x * cos(theta);
    double yf = y + vehicle_geometrics_.f2x * sin(theta);
    double x_inf = planning_scale_.xmin + vehicle_geometrics_.radius * 1.01;
    double x_sup = planning_scale_.xmax - vehicle_geometrics_.radius * 1.01;
    double y_inf = planning_scale_.ymin + vehicle_geometrics_.radius * 1.01;
    double y_sup = planning_scale_.ymax - vehicle_geometrics_.radius * 1.01;
    if (xr > x_sup || xr < x_inf ||
        xf > x_sup || xf < x_inf ||
        yr > y_sup || yr < y_inf ||
        yf > y_sup || yf < y_inf) {
        return false;
    }
    math::Vec2d idx_r = calc_xy_index(math::Vec2d(xr,yr));
    math::Vec2d idx_f = calc_xy_index(math::Vec2d(xf,yf));
    if (costmap_.at<uchar>(idx_r.y(),idx_r.x())==255 or costmap_.at<uchar>(idx_f.y(),idx_f.x())==255){
        return false;
    }
    return true;
}

bool IsCross(vector<math::Vec2d> vehicle, vector<math::Vec2d> obstacle) {
    bool IsCross = false;
    for (unsigned int i = 0; i < vehicle.size() - 1; i++) {
        if (checkObj_linev(vehicle[i], vehicle[i + 1], obstacle)) {
            IsCross = true;
            break;
        }
    }
    return IsCross;
}
